core/synchronizers.py

```diff
@@ -28,16 +44,121 @@ def test_connection(self) -> None:
+    User or contact are found based on the oidc_id or email.
+    If the oidc_id is provided, the email is ignored.
+    If the oidc_id is not provided, the email is used to find the user or contact but only select
+     users without oidc_id
+    If not match is found, function return None
+    If a match is found and updated, the function return the updated entity
+    """
+    if not oidc_id and not email:
+        return None
+    if not oidc_id:
+        try:
+            entity = user_or_contact_model.objects.get(email=email)
+        except user_or_contact_model.DoesNotExist:
+            return None
+        except user_or_contact_model.MultipleObjectsReturned as e:
+            raise InconsistentSynchronizerStateException(
+                f"Multiple {user_or_contact_model.__name__} found for this email {email}",
+            ) from e
+        # we only update users based on email if their oidc_id is not set
+        if entity.oidc_id:
+            raise InconsistentSynchronizerStateException(
+                f"OIDC don't match for {user_or_contact_model.__name__} with email {email} {oidc_id} expected while {entity.oidc_id} found in daisy"
+            )
+    else:
+        try:
+            entity = user_or_contact_model.objects.get(oidc_id=oidc_id)
+        except user_or_contact_model.DoesNotExist:
+            return None
+    update_and_save_entity(entity, info)
+    entity.save()
+    return entity
+
+
+def update_user(
+    user_info: Dict[str, str],
+    oidc_id: Optional[str] = None,
+    email: Optional[str] = None,
+) -> Optional[User]:
+    """
+    Update the user record based on the information provided
+    Returns True if a user for this oidc_id or email was found and updated, False otherwise
+    """
+    return update_user_or_contact(User, user_info, oidc_id, email)
+
+
+def update_contact(
+    contact_info: Dict[str, str],
+    oidc_id: Optional[str] = None,
+    email: Optional[str] = None,
+) -> Optional[Contact]:
+    """
+    Update the contact record based on the information provided
+    Returns True if a contact for this oidc_id was found and updated, False otherwise
+    """
+    return update_user_or_contact(Contact, contact_info, oidc_id, email)
+
+
+def create_contact(contact_dict):
+    contact = Contact()
+    update_and_save_entity(contact, contact_dict)
+    contact_type, _ = ContactType.objects.get_or_create(name="Other")
+    partner, _ = Partner.objects.get_or_create(
+        acronym="Imported from Keycloak", name="Imported from Keycloak"
+    )
+    contact.type = contact_type
+    contact.save()
+    contact.partners.add(partner)
+    contact.save()
+    return contact
+
+
+def update_and_save_entity(entity, info):
+    entity.last_name = info.get("last_name", entity.last_name)
+    entity.first_name = info.get("first_name", entity.first_name)
+    entity.email = info.get("email", entity.email)
+    entity.oidc_id = info.get("id")
```

Comment:
The entity is not saved here, do you think it is feasible to rename the method?

Reply:
thanks for spotting this, will fix

Reply:
I renamed the function and added docstring

Reply:
thanks